### PR TITLE
[RFC] ASoC: SOF: add abstraction for IPC, firmware and trace

### PR DESCRIPTION
**Pull request by plbossart**
In preparation for ICPv4 support, where we could have two types of firmware and IPC used on the same platform, this PR suggests a more granular split of the ops structure. The intent is for example that there is more than one IPC implementation.

This is only done for x84 platforms to get feedback for now.

### CODE REVIEW

**Comment by plbossart** [comment]
This seems to start to work for me. By default this uses the standard IPC and files.

with additional options I can switch the IPC type:
options snd-sof-pci fw_path="intel/cavs"
options snd-sof-pci ipc_type=1

````
[   35.323974] sof-audio-pci-intel-cnl 0000:00:1f.3: overriding default IPC 0 to requested 1
[   35.323977] sof-audio-pci-intel-cnl 0000:00:1f.3: Module parameter used, changed fw path to intel/cavs
[   35.323981] sof-audio-pci-intel-cnl 0000:00:1f.3: error: missing mandatory ops
````

we'd need to update the Intel private branch with the cavs code, but this seems to be a promising path for coexistence between IPC types.

File: include/sound/sof.h [hunk 1]
```diff
@@ -90,6 +90,8 @@ struct snd_sof_pdata {
 	const struct snd_soc_acpi_mach *machine;
 
 	void *hw_pdata;
+
+	enum sof_ipc_types ipc_type;
```
Comment (ranj063) [inline include/sound/sof.h]: Looks great @plbossart . The only opens I have are that the dsp ops partitioned in the first patch need to also be made of size SOF_IPC_COUNT, right ? And the way you've defined the cavs IPC, it can't use the SOF firmware. Is that ok? What happens when SOF firmware starts supporting IPC4

File: sound/soc/sof/ops.h [hunk 2]
```diff
@@ -21,6 +21,15 @@
 #define sof_ops(sdev) \
 	((sdev)->pdata->desc->ops)
 
+#define sof_ipc_ops(sdev) \
```
Comment (ranj063) [inline sound/soc/sof/ops.h]: Should we rename this to sof_dsp_ipc_ops? And all the rest too for consistency? Because my PR also adds the top-level ipc ops

File: sound/soc/sof/sof-priv.h [hunk 3]
```diff
@@ -118,6 +118,52 @@ struct snd_sof_platform_stream_params {
...
 };
 
+struct snd_sof_dsp_ipc_ops {
+	/* doorbell */
+	irqreturn_t (*irq_handler)(int irq, void *context); /* optional */
+	irqreturn_t (*irq_thread)(int irq, void *context); /* optional */
+
+	/* ipc */
+	int (*send_msg)(struct snd_sof_dev *sof_dev,
+			struct snd_sof_ipc_msg *msg); /* mandatory */
```
Comment (ujfalusi) [inline sound/soc/sof/sof-priv.h:131]: The `msg` parameter for `ipc_msg_data()` / `send_msg()` is always exclusively `&sdev->ipc->msg`
The RX (reply or notification) also uses the `sdev->ipc->msg`, I don't think that there is even a slight possibility to give 'custom' `struct snd_sof_ipc_msg` as we need to have serialization among messages.
Would it make sense to drop the `msg` around this move?

Comment (ujfalusi) [reply sound/soc/sof/sof-priv.h:131]: We also need another layer of ops on top for core IPC level handling, or if I think about it we might not need them at all as the functions calling tx_message, set_get_data and friends have been moved to IPC specific ops, thus they can just happily call IPC specific functions?

Fwiw I have mostly implemented these for IPC3:
```c
struct ipc_ops {
 	const struct ipc_tplg_ops *tplg;
 	const struct ipc_pm_ops *pm;
 	const struct ipc_pcm_ops *pcm;
+
+	int (*tx_msg)(struct snd_sof_dev *sdev, void *msg_data, size_t msg_bytes,
+		      void *reply_data, size_t reply_bytes, bool no_pm);
+	void (*rx_msg)(struct snd_sof_dev *sdev);
+	int (*set_data)(struct snd_sof_dev *sdev, void *data, size_t data_bytes);
+	int (*get_data)(struct snd_sof_dev *sdev, void *data, size_t data_bytes);
 };
```

File: sound/soc/sof/sof-priv.h [hunk 4]
```diff
@@ -118,6 +118,52 @@ struct snd_sof_platform_stream_params {
...
+	 */
+	int (*fw_ready)(struct snd_sof_dev *sdev, u32 msg_id); /* mandatory */
+
+	/* host side configuration of the stream's data offset in stream mailbox area */
+	int (*set_stream_data_offset)(struct snd_sof_dev *sdev,
+				      struct snd_pcm_substream *substream,
+				      size_t posn_offset); /* optional */
+
+	/* host read DSP stream data */
+	int (*ipc_msg_data)(struct snd_sof_dev *sdev,
```
Comment (ujfalusi) [inline sound/soc/sof/sof-priv.h:145]: get_msg_data might be better name to indicate the direction?

File: sound/soc/sof/sof-priv.h [hunk 5]
```diff
@@ -118,6 +118,52 @@ struct snd_sof_platform_stream_params {
...
+	int (*load_module)(struct snd_sof_dev *sof_dev,
+			   struct snd_sof_mod_hdr *hdr); /* optional */
+	/* parse platform specific extended manifest, optional */
+	int (*parse_platform_ext_manifest)(struct snd_sof_dev *sof_dev,
+					   const struct sof_ext_man_elem_header *hdr);
+};
+
+struct snd_sof_dsp_trace_ops {
+	int (*trace_init)(struct snd_sof_dev *sdev,
+			  struct sof_ipc_dma_trace_params_ext *dtrace_params); /* optional */
```
Comment (ujfalusi) [inline sound/soc/sof/sof-priv.h:164]: `struct sof_ipc_dma_trace_params_ext` is an IPC3 message...

File: sound/soc/sof/sof-priv.h [hunk 6]
```diff
@@ -118,6 +118,52 @@ struct snd_sof_platform_stream_params {
...
+	int (*trace_release)(struct snd_sof_dev *sdev); /* optional */
+	int (*trace_trigger)(struct snd_sof_dev *sdev,
+			     int cmd); /* optional */
+};
```
Comment (ujfalusi) [inline sound/soc/sof/sof-priv.h:168]: I would be tempted to move all dtrace related options under a struct, but it is unrelated.

File: include/sound/soc-acpi.h [hunk 7]
```diff
@@ -142,7 +142,6 @@ struct snd_soc_acpi_link_adr {
  * audio codecs whose presence if checked with ACPI
  * @pdata: intended for platform data or machine specific-ops. This structure
  *  is not constant since this field may be updated at run-time
- * @sof_fw_filename: Sound Open Firmware file name, if enabled
```
Comment (ujfalusi) [inline include/sound/soc-acpi.h:145]: I guess the 'if enabled' meant 'if set'?

The commit message is a bit misleading as we were using the `sof_fw_filename` whenever it was set, I think it should read that we used the _same_ firmware file name in `sof_fw_filename` as the default (pdata->desc->default_fw_filename), removing the `sof_fw_filename` will result using the default_fw_filename , which is the same for these platforms?

File: include/sound/sof.h [hunk 8]
```diff
@@ -47,6 +47,13 @@ enum sof_dsp_power_states {
 	SOF_DSP_PM_D3,
 };
 
+/* Definitions for multiple IPCs */
+enum sof_ipc_types {
+	SOF_IPC,
+	SOF_INTEL_CAVS,
+	SOF_IPC_COUNT
```
Comment (ujfalusi) [inline include/sound/sof.h]: SOF_IPC_TYPE_COUNT might be better to understand for the fir look? Or to the extreme: SOF_SUPPORTED_IPC_TYPE_COUNT.

Comment (ujfalusi) [reply include/sound/sof.h]: Please also add a semicolon after the last entry.

File: sound/soc/sof/sof-pci-dev.c [hunk 9]
```diff
@@ -31,6 +31,10 @@ static int sof_pci_debug;
 module_param_named(sof_pci_debug, sof_pci_debug, int, 0444);
 MODULE_PARM_DESC(sof_pci_debug, "SOF PCI debug options (0x0 all off)");
 
+static int sof_pci_ipc_type = SOF_IPC;
```
Comment (ujfalusi) [inline sound/soc/sof/sof-pci-dev.c]: I think this should be `sof_pci_ipc_type = SOF_IPC_COUNT;`

File: sound/soc/sof/sof-pci-dev.c [hunk 10]
```diff
@@ -152,6 +156,23 @@ int sof_pci_probe(struct pci_dev *pci, const struct pci_device_id *pci_id)
 	sof_pdata->dev = dev;
 	sof_pdata->fw_filename = desc->default_fw_filename;
 
+	sof_pdata->ipc_type = desc->ipc_default;
+
+	if (sof_pci_ipc_type != desc->ipc_default) {
```
Comment (ujfalusi) [inline sound/soc/sof/sof-pci-dev.c]: While this is true, it is not correct.
if the platform by default uses INTEL_CAVS (desc->ipc_default == SOF_INTEL_CAVS) and you don't specify any options to change that then you will override it to IPC_SOF.

I think the logic should be something like:
```c
if (sof_pci_ipc_type == SOF_IPC_COUNT) {
	sof_pdata->ipc_type = desc->ipc_default;
}else {
	if (sof_pci_ipc_type < 0 || sof_pci_ipc_type >= SOF_IPC_COUNT) {
		dev_err(dev, "invalid request value %d\n", sof_pci_ipc_type);
		return -EINVAL;
	}
	if (!(BIT(sof_pci_ipc_type) & desc->ipc_supported_mask)) {
		dev_err(dev, "invalid request value %d, supported mask is %#x\n",
			sof_pci_ipc_type, desc->ipc_supported_mask);
		return -EINVAL;
	}
	sof_pdata->ipc_type = sof_pci_ipc_type;
	dev_info(dev, "overriding default IPC %d to requested %d\n", desc->ipc_default, sof_pci_ipc_type);
}
```

**Review by ujfalusi — CHANGES_REQUESTED** [review]
@plbossart, looks nice and finally it ticked in place in my brain that this is like a chicken-egg problem.. The used IPC depends on the firmware we load, but we need to load different firmware, so we first need to be able to select which way we are going.

I have some comments here and there.

File: sound/soc/sof/ops.h (hunk 2 above)
Comment (plbossart) [reply sound/soc/sof/ops.h]: sure, this name is better. will change.

File: sound/soc/sof/sof-priv.h (hunk 3 above)
Comment (plbossart) [reply sound/soc/sof/sof-priv.h:131]: > The `msg` parameter for `ipc_msg_data()` / `send_msg()` is always exclusively `&sdev->ipc->msg` The RX (reply or notification) also uses the `sdev->ipc->msg`, I don't think that there is even a slight possibility to give 'custom' `struct snd_sof_ipc_msg` as we need to have serialization among messages. Would it make sense to drop the `msg` around this move?

that would need to be a separate change. We can do this before this PR, or after this PR. I don't want to add any functionality while splitting the existing code in separate ops.

Comment (plbossart) [reply sound/soc/sof/sof-priv.h:131]: > We also need another layer of ops on top for core IPC level handling, or if I think about it we might not need them at all as the functions calling tx_message, set_get_data and friends have been moved to IPC specific ops, thus they can just happily call IPC specific functions?
> 
> Fwiw I have mostly implemented these for IPC3:
> 
> ```c
> struct ipc_ops {
>  	const struct ipc_tplg_ops *tplg;
>  	const struct ipc_pm_ops *pm;
>  	const struct ipc_pcm_ops *pcm;
> +
> +	int (*tx_msg)(struct snd_sof_dev *sdev, void *msg_data, size_t msg_bytes,
> +		      void *reply_data, size_t reply_bytes, bool no_pm);
> +	void (*rx_msg)(struct snd_sof_dev *sdev);
> +	int (*set_data)(struct snd_sof_dev *sdev, void *data, size_t data_bytes);
> +	int (*get_data)(struct snd_sof_dev *sdev, void *data, size_t data_bytes);
>  };
> ```
not following this comment at all. What PR are you referring to @ujfalusi ?

File: include/sound/sof.h (hunk 1 above)
Comment (plbossart) [reply include/sound/sof.h]: > The only opens I have are that the dsp ops partitioned in the first patch need to also be made of size SOF_IPC_COUNT, right ? 

Not following this one @ranj063, can you explain what you are referring to?

> And the way you've defined the cavs IPC, it can't use the SOF firmware. Is that ok? What happens when SOF firmware starts supporting IPC4

 The idea for this PR is that for each platform there is a default IPC version, which can be overridden manually by developers. If there are multiple implementations of the IPCv4, I think again we'll have to select a default path and manually override if we want to switch between implementations. For example, instead of 'intel/avs' the path would be 'intel/sof-ipcv4'. The file names and topologies would have to be the same though.

File: sound/soc/sof/sof-priv.h (hunk 3 above)
Comment (ujfalusi) [reply sound/soc/sof/sof-priv.h:131]: @plbossart, I don't have a PR as my wip is based on several PRs in flight and have daily conflicts when rebasing on the updated PRs.

The send_msg() here is the platform IPC message sending, which can differ between IPC versions.
In top level SOF we have the user facing tx_message() variants for IPC3 and IPC4, they are calling this platform callback, but they do IPC specific lifting as well.
So, the question is: do we want ops for that level or we are calling the functions from already IPC dependent code.
I have a generic tx_msg() for ipc3/4 and set_get_data() for ipc3 implemented. We can have some shared code left in ipc.c, but most of the content there atm is realy ipc3 only.

Comment (plbossart) [reply sound/soc/sof/sof-priv.h:131]: The intent of the ops I added was to cover the low-level combination of hardware specific + IPC type specific code.

The interrupt handling falls in this category, since we are using different registers for IPC between hardware generations.

I am not completely sure if there is merit is abstracting in sof_dsp_ipc_ops() any functionality that is NOT hardware specific. That should be handled in the higher level code of ipcv3/ipcv4.

So with that, I am not sure if the partitioning I chose is actually relevant. I don't think there's any issue with 
- irq_handler
- irq_thread
- fw_ready

I am not sure about 
- send_msg
- ipc_msg_data

I am even less sure about 
- ipc_dump. I don't know if we use hardware-specific resources here.
- set_stream_data_offset. This is a bit inconsistent because I've left get_mailbox_offset() and get_window_offset() out.

In other words, my main ask is to review what needs to be in these sof_dsp_ipc_ops structures, and what is handled at a higher level.

File: include/sound/sof.h (hunk 1 above)
Comment (ranj063) [reply include/sound/sof.h]: > Not following this one @ranj063, can you explain what you are referring to?
> 
> >

https://github.com/thesofproject/linux/blob/9ea6ec88b75bae1bb07a24fd0d254da4a76f97c7/sound/soc/sof/intel/cnl.c#L246

I was wondering if the array above needs to be of size SOF_IPC_COUNT too

Comment (plbossart) [reply include/sound/sof.h]: yes, nice catch indeed. It only works today because I only added the default index 0, but that wouldn't work if I want to define only INTEL_CAVS, as planned for SKL/KBL.

Thanks for spotting this @ranj063

Comment (plbossart) [reply include/sound/sof.h]: fixed

File: sound/soc/sof/sof-pci-dev.c (hunk 9 above)
Comment (plbossart) [reply sound/soc/sof/sof-pci-dev.c]: nope, that would cause out-of-bounds access.

File: sound/soc/sof/ops.h (hunk 2 above)
Comment (plbossart) [reply sound/soc/sof/ops.h]: done

File: sound/soc/sof/sof-priv.h (hunk 4 above)
Comment (plbossart) [reply sound/soc/sof/sof-priv.h:145]: no rename or functionality change in this PR

File: include/sound/sof.h (hunk 8 above)
Comment (plbossart) [reply include/sound/sof.h]: > SOF_IPC_TYPE_COUNT might be better to understand for the fir look? Or to the extreme: SOF_SUPPORTED_IPC_TYPE_COUNT.

I will change to SOF_IPC_TYPE_COUNT.

Comment (plbossart) [reply include/sound/sof.h]: > Please also add a semicolon after the last entry.

err, no. we absolutely do not want any future addition after this line.

File: include/sound/soc-acpi.h (hunk 7 above)
Comment (plbossart) [reply include/sound/soc-acpi.h:145]: reworded as 

"  We've been using a default firmware name for each PCI/ACPI/OF platform
    for a while. The machine-specific sof_fw_filename is in practice not
    different from the default, and newer devices don't set this field, so
    let's remove the redundant definitions.
"

File: sound/soc/sof/sof-pci-dev.c (hunk 10 above)
Comment (plbossart) [reply sound/soc/sof/sof-pci-dev.c]: Now I get what you were saying @ujfalusi 

I took your suggestion but used -1 as default for consistency with other kernel module parameters.

File: sound/soc/sof/sof-priv.h (hunk 5 above)
Comment (plbossart) [reply sound/soc/sof/sof-priv.h:164]: that's not so good. I don't know how to fix this one.

Comment (ranj063) [reply sound/soc/sof/sof-priv.h:164]: void *params maybe? then we can set struct sof_ipc_dma_trace_params_ext *dtrace_params = params in hda-trace.c hda_dsp_trace_init()

File: sound/soc/sof/ops.h [hunk 11]
```diff
@@ -21,6 +21,15 @@
 #define sof_ops(sdev) \
 	((sdev)->pdata->desc->ops)
 
+#define sof_dsp_ipc_ops(sdev) \
+	((sdev)->pdata->desc->ops->ipc_ops)
```
Comment (ranj063) [inline sound/soc/sof/ops.h]: can we drop the _ops suffix from ipc_ops, fw_ops and trace_ops? ((sdev)->pdata->desc->ops->ipc) already has ops in it

File: sound/soc/sof/sof-priv.h (hunk 5 above)
Comment (ujfalusi) [reply sound/soc/sof/sof-priv.h:164]: The trace_init passes the dtrace_params because
Intel HDA needs to update the stream_tag in the message.
AMD ACP needs to override the buffer.phy_addr (it also provides a stream_tag, but afaik it is not used)
So in essence it is set up the platform side and also to feed back information. How did we do this for the audio streams? They also need the stream_tag, no?

My other take is that the dtrace is SOF (IPC3) specific, IPC4 uses different trace, they also have DMA backed debugging feature, but not sure how that compares to our dtrace...

File: sound/soc/sof/sof-priv.h (hunk 3 above)
Comment (ujfalusi) [reply sound/soc/sof/sof-priv.h:131]: `send_msg` iand `ipc_msg_data` s really IPC specific, the sending of a message is done differently for IPC3 compared to IPC4, just enough that it is cleaner to have different ops.
`ipc_dump` is more like platform than IPC related by the implementations.
`set_stream_data_offset`, is probably again fall to platform along with the get_mailbox_offset() and window_offset()?

About the `snd_sof_ipc_set_get_comp_data()`, `sof_ipc_tx_message()` and `sof_ipc_tx_message_no_pm()`, we don't want them to be a generic interface, but rename them like s/ipc/ipc3 and have ipc4 variants (we already have `ipc4_tx_message()`) when needed, right?
Related things, like `tx_wait_done()` and `ipc_log_header()` are IPC specific and not that reusable, unless we move the wait down to platform level, within `snd_sof_dsp_send_msg()` callback.

Comment (plbossart) [reply sound/soc/sof/sof-priv.h:131]: Sorry @ujfalusi you've lost me. I am not that familiar with the IPC and I don't know what is remark/question/suggestion/recommendation.

It'd be more productive for me if you suggested a code diff or patch on what you think would make sense.

File: sound/soc/sof/core.c [hunk 12]
```diff
@@ -357,6 +357,14 @@ int snd_sof_device_probe(struct device *dev, struct snd_sof_pdata *plat_data)
 	sdev->first_boot = true;
 	dev_set_drvdata(dev, sdev);
 
+	/* check IPC support */
+	if (!(BIT(plat_data->ipc_type) & plat_data->desc->ipc_supported_mask)) {
+		dev_err(dev, "ipc_type %d is not supported on this platform, mask is %#x\n",
+			plat_data->ipc_type, plat_data->desc->ipc_supported_mask);
+		return -EINVAL;
```
Comment (bardliao) [inline sound/soc/sof/core.c:364]: Will this really happen? We already test sof_pci_ipc_type on sof-pci-dev.c.

File: sound/soc/sof/sof-priv.h (hunk 3 above)
Comment (ujfalusi) [reply sound/soc/sof/sof-priv.h:131]: @plbossart, OK, I'll have a draft tomorrow. I have the IPC3 now working, but I'm not happy with the code duplication between IPC versions.

File: sound/soc/sof/core.c (hunk 12 above)
Comment (plbossart) [reply sound/soc/sof/core.c:364]: Not strictly necessary but a good paranoia check. I wanted to double-check for the other cases (ACPI and OF) that we didn't miss a platform.

File: include/sound/sof.h [hunk 13]
```diff
@@ -47,6 +47,13 @@ enum sof_dsp_power_states {
 	SOF_DSP_PM_D3,
 };
 
+/* Definitions for multiple IPCs */
+enum sof_ipc_types {
```
Comment (ujfalusi) [inline include/sound/sof.h]: I think this should not be plural: `enum sof_ipc_type`

File: sound/soc/sof/intel/byt.c [hunk 14]
```diff
@@ -215,8 +215,37 @@ static int byt_acpi_probe(struct snd_sof_dev *sdev)
...
+		.send_msg	= atom_send_msg,
+		.fw_ready	= sof_fw_ready,
+
+		.ipc_msg_data	= sof_ipc_msg_data,
+		.set_stream_data_offset = sof_set_stream_data_offset,
+	},
+};
+
+static const struct snd_sof_dsp_fw_ops sof_byt_fw_ops[SOF_IPC_TYPES_COUNT] = {
+	[SOF_IPC] {
```
Comment (ujfalusi) [inline sound/soc/sof/intel/byt.c]: missing `=`

**Review by ujfalusi — CHANGES_REQUESTED** [review]
iMX and mediatek conversion is missing.

**Comment by plbossart** [comment]
@ujfalusi I rebased on top of PR https://github.com/thesofproject/linux/pull/3382 so that only the last 2 commits are related to the ipc4 abstractions. I am not sure if these abstractions make any sense, or if they duplicate what you've been doing. Care to comment or point me to the 'right' branch?

File: sound/soc/sof/intel/byt.c (hunk 14 above)
Comment (plbossart) [reply sound/soc/sof/intel/byt.c]: fixed

File: include/sound/sof.h (hunk 13 above)
Comment (plbossart) [reply include/sound/sof.h]: done

**Comment by plbossart** [comment]
> iMX and mediatek conversion is missing.

I'll do it when we agree on Intel stuff :-)
